dbms/src/Common/StringUtils/StringUtils.h

```diff
@@ -149,3 +149,99 @@ inline bool equalsCaseInsensitive(char a, char b)
 {
     return a == b || (isAlphaASCII(a) && alternateCaseIfAlphaASCII(a) == b);
 }
+
+inline bool isPunctuation(char c)
+{
+    return (c >= 0x21 && c <= 0x2F) || (c >= 0x3A && c <= 0x40) || (c >= 0x5B && c <= 0x60) || (c >= 0x7B && c <= 0x7E);
+}
+
+// match at lease x digits
+inline std::tuple<std::string, std::string, bool> digit(const std::string & str, int x)
+{
+    for (size_t i = 0, times = x; i < times && i < str.length() && isdigit(str[i]); i++, x--)
```

Comment:
Maybe we should get str length in advance so that we can avoid useless function calls

Reply:
> Maybe we should get str length in advance so that we can avoid useless function calls

IMO, compiler will do this for us.

Reply:
> 

I agree